Added support for voidtools everything DB

cobyge opened the pull request:
Inspired by #505, I remembered I had some code lying around to parse the database of Voidtools Everything, very similar to mlocate/plocate, but for Windows.

I updated the code and added it to the codebase.  
Because Everything is closed source, this is completely based off of reverse-engineering the code, and I haven't found any reference implementation on the internet to help (AFAIK this is the only parser), so this is all based off of my (not too great) reversing skills.  
I've tested this on ~10 random database files I had lying around, from multiple computers, all of them have given exactly the same exact results as Everything itself (checked by exporting to CSV and comparing md5sums).  
It should support any DB created since 2017, and if given a broken file, I'm willing to add support for earlier versions as well.

All comments are mine, written while reversing the code.  

This is relatively slow code (takes 4.5 seconds for a DB with 126828 files), 
I have a version written in Rust which is 22 times faster, and if that's something you are interested in, then I'm happy to try creating bindings with Py03.

cobyge commented:
I've now added support for all filesystem types (Currently NTFS/REFS/EFU/Folder), along with tests for each.

Next step is to add support for more versions (Everything 1.5.0alpha currently uses version 1.7.49 and also supports FAT, network drives, and network indexes)

Horofic commented:
Really cool PR! Since this is another big one, please give it some time for us to do the review :). Stay tuned!

cobyge commented:
Hey, thanks for the review.
I updated the code according to your request, and I've also added support for a previous version of Everything, in order to what supporting multiple versions might look like.

The only request I haven't worked on yet is the request regarding using dissect.cstruct. I'll have to think a bit about how to implement it, because of differences between structs for multiple versions.

I'd be happy to hear thoughts about how I handled different versions in the code (I'm not quite happy about with the version handling).

cobyge commented:
Hey!
It's been a while, but I finally got around to rewriting this with a simpler (better?) implementation, using cstruct.

I haven't been following any project changes for a while, so if I need to change anything let me know (the tests pass 🤷).

I'm resolving the previous comments as they are no longer relevant.
Hope we can get this pushed sometime soon 😄

cobyge commented:
Thanks for the review.
I'm vacationing the sea for the next couple of weeks, so hopefully I'll get to this right after that

EinatFox commented:
Hi @cobyge , hope you are doing well. Just checking in to know if you plan to process this review? Thanks

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,546 @@
+                trunc_from_prev = self.read_byte_or_4()
+                if trunc_from_prev > len(temp_buf):
+                    raise ValueError(f"Error while parsing file name {trunc_from_prev} > {len(temp_buf)}")
+                temp_buf = temp_buf[: len(temp_buf) - trunc_from_prev]
```
In between folder and file parsing, I reset tmp_buf
`temp_buf = b""` (line 455).

Either way, the first file _must_ have a name, so the first file in the list will always enter into `if new_byte_count` and be overriden

cobyge commented:
Hey, I actually totally forgot about this, sorry about that 😅

Going over the comments now and fixing them.
Something I just noticed when rebuilding this is that [packaging](https://github.com/pypa/packaging) is a new dependency that is no used so far. Would it be okay to add it as a dependency? Its an official pypy package if that helps

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,546 @@
+    uint8_t version_minor;
+    uint8_t version_major;
+
+    // Flags
+    uint32_t flag_has_file_size:1;
+    uint32_t flag_has_date_created:1;
+    uint32_t flag_has_date_modified:1;
+    uint32_t flag_has_date_accessed:1;
+    uint32_t flag_has_attributes:1;
+    uint32_t flag_has_folder_size:1;
```
I changed my code to use cstruct flags, but it then caused me to have to put:
```python
if self.header.flags.value && self.header.flags.has_date_modified: ...
```
Instead of 
```python
if self.header.flag_has_folder_size: ...
```
everywhere, and of course everything is untyped. I don't think that it's worth using the flag in this case.
If you want I'll add it though

cobyge commented on dissect/target/plugins/os/windows/everything/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from .everything import EverythingPlugin
+
+__all__ = ["EverythingPlugin"]
```
I used the new strategy from #788, I'd appreciate it if you make sure I did all that needs to be done

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,546 @@
+            try:
+                fs_type = EverythingFSType(filesystem_header.type)
+            except ValueError:
+                raise ValueError(f"Unknown FS type {filesystem_header.type}") from None
```
I personally prefer `raise from None` in this situation, there is a huge readability difference when we want to swallow errors. (Either way I've removed the `from None`, let me know if you want me to put it back)
Here is an example:
```python
from enum import IntEnum
class A(IntEnum):
    T = 0
```


```python
try:
    A(1)
except ValueError:
    raise ValueError(f"Invalid type") from None

Traceback (most recent call last):
  File "/Users/coby/p/git/dissect.target/a.py", line 16, in <module>
    raise ValueError(f"Invalid type") from None
ValueError: Invalid type
```

```python
try:
    A(1)
except ValueError:
    raise ValueError(f"Invalid type")

Traceback (most recent call last):
  File "/Users/coby/p/git/dissect.target/a.py", line 8, in <module>
    A(1)
    ~^^^
  File "/Users/coby/.local/share/uv/python/cpython-3.13.0-macos-aarch64-none/lib/python3.13/enum.py", line 722, in __call__
    return cls.__new__(cls, value)
           ~~~~~~~~~~~^^^^^^^^^^^^
  File "/Users/coby/.local/share/uv/python/cpython-3.13.0-macos-aarch64-none/lib/python3.13/enum.py", line 1189, in __new__
    raise ve_exc
ValueError: 1 is not a valid A

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/coby/p/git/dissect.target/a.py", line 10, in <module>
    raise ValueError(f"Invalid type")
ValueError: Invalid type
```

cobyge commented:
Finished with the comments, I left some comments unresolved that I need your response/approval for

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,551 @@
+
+            if fs_type in [EverythingFSType.NTFS, EverythingFSType.REFS]:
+                if fs_type == EverythingFSType.NTFS:
+                    c_fs = self.filesystems_cstruct.ntfs_header(self.fh)
```
I'm not sure what you mean, can you explain?

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,551 @@
+log = logging.getLogger(__name__)
+
+
+class EverythingVarInt(BaseType):
```
Can you give me a summary of what should be changed? I don't see of any way for me to know what is relevant

cobyge commented:
Hey, I've gone over your comments again, fixed everything except for the ones I left responses on.

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,542 @@
+            folder.attributes = 16
+
+            # The yield can't happen here, because we can only call resolve_path once we finish this loop
+            if self.header.entry_attributes_flags.has_folder_size in self.header.entry_attributes_flags:
+                folder.size = c_everything_header.uint64_t(self.fh)
+            if self.header.entry_attributes_flags.has_date_created in self.header.entry_attributes_flags:
+                folder.date_created = c_everything_header.uint64_t(self.fh)
+            if self.header.entry_attributes_flags.has_date_modified in self.header.entry_attributes_flags:
+                folder.date_modified = c_everything_header.uint64_t(self.fh)
+            if self.header.entry_attributes_flags.has_date_accessed in self.header.entry_attributes_flags:
+                folder.date_accessed = c_everything_header.uint64_t(self.fh)
+            if self.header.entry_attributes_flags.has_attributes in self.header.entry_attributes_flags:
+                folder.attributes = c_everything_header.uint32_t(self.fh)
```
Done

Schamper commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,551 @@
+    def parse_files(self, folder_list: list[EverythingIndexObj]) -> Iterator[EverythingF]:
+        temp_buf = b""
+        for _ in range(self.header.number_of_files):
+            # See comment above for explanation of this loop
```
Which loop is this comment referring to?

Schamper commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,551 @@
+            # and can be fixed by reading and discarding a single byte here.
+
+            file_name = folder_list[parent_index].resolve_path(folder_list)
+            if new_byte_count := read_byte_or_4(self.fh):
+                trunc_from_prev = read_byte_or_4(self.fh)
+                if trunc_from_prev > len(temp_buf):
+                    raise ValueError(f"Error while parsing file name {trunc_from_prev} > {len(temp_buf)}")
+                temp_buf = temp_buf[: len(temp_buf) - trunc_from_prev]
+            temp_buf += self.fh.read(new_byte_count)
```
I think this is the same reading loop as before, maybe move this to a function in the global scope called `read_truncated_name` or something?

cobyge commented on dissect/target/plugins/os/windows/everything/__init__.py:
Let me know if you want me to move it

Schamper commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,452 @@
+            self.filesystem_list.append((fs_type, filesystem_header, header))
+
+    def __iter__(self) -> Iterator[EverythingF]:
+        self.fh.seek(self._start_seek, io.SEEK_SET)
```
```suggestion
        self.fh.seek(self._start_seek)
```
`SEEK_SET` is already the default.

Schamper reviewed (changes requested):
And some comments on the previous threads.

cobyge commented on dissect/target/plugins/os/windows/everything/everything_types.py:
```diff
@@ -0,0 +1,452 @@
+
+
+@dataclasses.dataclass
+class EverythingF:
```
Nothing, renamed to `Record` instead

Schamper commented:
> If you want me to do this anyways, that's fine, just let me know

No it's fine, let's keep it this way. One small change and this should be good to merge. I'll put the outstanding comment about moving it to a different spot in a new ticket for future consideration.

cobyge commented on dissect/target/plugins/os/windows/everything/parser.py:
```diff
@@ -0,0 +1,452 @@
+        return f"{self.header} - Filesystems{self.filesystem_list}"
+
+    def __parse_filesystems(self) -> None:
+        self.filesystem_list = []
```
Done 😄

Schamper reviewed (approved):
@Miauwkeru pinging.
